lib/std/log.zig

```diff
@@ -0,0 +1,74 @@
+const std = @import("std.zig");
+const debug = std.debug;
+const builtin = std.builtin;
+const root = @import("root");
+
+pub const Level = enum {
+    /// helpful information for debugging.
+    Info,
+    /// can't be sure something is wrong, but it is probably worth investigating
+    Warning,
+    /// a bug in the application has been detected, but it is recoverable.
+    Error,
+    /// a bug in the application has been detected, and the application must
+    /// terminate now.
+    Fatal,
+};
+
+/// The default log level is based on build mode. Note that in ReleaseSmall
+/// builds logging is totally disabled (log.fatal still causes a panic though).
+pub const default_level: Level = switch (builtin.mode) {
+    .Debug => .Info,
+    .ReleaseSafe => .Warning,
+    .ReleaseFast => .Error,
+    .ReleaseSmall => .Fatal,
+};
+
+pub const level: Level = if (@hasDecl(root, "log_level")) root.log_level else default_level;
+
+fn log(message_level: Level, comptime format: []const u8, args: var) void {
```

Comment:
Should the `message_level` arg be made comptime, to avoid code generation in release modes? Or is the zig compiler able to figure this out implicitly?

Reply:
certainly couldn't hurt

Reply:
As a real-world example, Unreal Engine uses macros to have both a comptime minimum log level and a runtime-controlled log level, so you can e.g. compile out verbose log statements but turn on/off info log statements without a recompile.

Reply:
@SpexGuy I think the way #2586 is going is to keep the default logging only configurable at comptime. However, runtime control can be gained if desired by defining `root.log` which receives the log level any may implement comptime or runtime configuration.